cvmfs/cache_ram.cc

```diff
@@ -92,22 +142,30 @@ int64_t RamCacheManager::Pread(
   uint64_t size,
   uint64_t offset)
 {
-  return -EIO;
+  ReadLockGuard guard(rwlock_);
+  if (!IsValid(fd)) {
+    LogCvmfs(kLogCache, kLogDebug, "bad fd %d on Pread", fd);
+    return -EBADF;
+  }
+  assert(open_fds_[fd].store);
+  perf::Inc(counters_.n_pread);
+  return open_fds_[fd].store->Read(open_fds_[fd].handle, buf, size, offset);
 }
 
 
 int RamCacheManager::Dup(int fd) {
-  ReadOnlyFd file_descriptor;
-  {
-    ReadLockGuard guard(rwlock_);
-    if (!IsValid(fd))
-      return -EBADF;
-    file_descriptor = open_fds_[fd];
-  }
-
-  // TODO(jblomer): increase link count in kv store
+  bool rc;
   WriteLockGuard guard(rwlock_);
-  return AddFd(file_descriptor);
+  if (!IsValid(fd)) {
+    LogCvmfs(kLogCache, kLogDebug, "bad fd %d on Dup", fd);
+    return -EBADF;
+  }
+  assert(open_fds_[fd].store);
+  rc = open_fds_[fd].store->IncRef(open_fds_[fd].handle);
+  assert(rc);
+  LogCvmfs(kLogCache, kLogDebug, "dup fd %d", fd);
+  perf::Inc(counters_.n_dup);
+  return AddFd(open_fds_[fd]);
```

Comment:
I think you'd need to check the return value of `AddFd` and decrease the reference count if it is unsuccessful.